### DISCUSSION TITLE
Layer-specific gcode for all prints

### DISCUSSION THREAD

**Discussion by lullaby37** [OrcaSlicer/OrcaSlicer#2225] [Q&A] [answered]
Hello,

I've been using OrcaSlicer for a while now and I really like it.
My DIY-printer running Klipper has a smart filament motion sensor that also triggers when the filament motion is not as expected.
I activate the sensor at the beginning of the print, the problem with that is that on the firt layer the false-positive activation of the sensor is too high.
So my idea was to somehow activate the sensor on layer 2 or 3.
No I know how to do that manually for a colour change, for example. But is there a function to create this automatically for each layer? That would be very helpful.
Thanks!

**Comment by foreachthing** [comment]
How do you "activate the sensor"?
You might be able to solve your problem in post. With Python you could grab your command and move it to layer x.

**Comment by lullaby37** [comment]
Thanks for the answer.
The sensor is switched off on boot and the following lines activates it/ switches it on:
`SET_FILAMENT_SENSOR SENSOR=SFS_T0 ENABLE=1` 
This line is part of my Custom Start gcode.

All I need is to find out how to run this command in the 2nd or 3rd layer. And I need to know how this is done automatically.
Manually I could do it by right clicking on a specific layer in the preview part of OrcaSlicer and then adding custom gcode.

**Comment by foreachthing** [comment]
How fit are you with Python 😉 (or any other language) ?
You basically need to loop over your gcode file until you find your line `SET_FILAMENT_SENSOR SENSOR=SFS_T0 ENABLE=1`. Then comment/remove it and loop until `;layer:2` (requires `;layer:[layer_num];` in machine settings "layer change gcode". 
That's it. If you want you can use parameters (like insert to layer x) with `argparse` to make it easier to change insertion point.

**Comment by lullaby37** [comment]
Thanks for your answer. I am not really good with Python and so I cannot really figure out what you mean.
The `SET_FILAMENT_SENSOR SENSOR=SFS_T0 ENABLE=1` is part of my start-gcode in Klipper, so I know where it can be found and I know how to remove it.
But the rest is double-dutch to me. Where do I have to add it in OrcaSlicer to automate that the sensor's state is set to "on" on layer two?
Sorry if that is a dumb question but I am a bit of a noob.

**Comment by foreachthing** [comment]
I'm far away from a computer rught now.
But, you can serach for prusa post processing scripts. Then you'll have to loop over your gcode file. Find your command, find layer 2 (or 3) and insert it again. Exit loop, save file. Done.
When I'm back I can help you further...

**Comment by mihi4** [comment]
Could you use the "change layer" g-code section in the machine settings?
something like (pseudocode) 
if layer_z == 2 SETFILAMENT_SENSOR SENSOR=SFS_T0 ENABLE=1

this code would be run on each layer change, but the sensor would only be activated on layer 2, on all others it would do nothing

**Reply by foreachthing** [reply]
Good idea!!
And even simpler.

**Reply by lullaby37** [reply]
Could you guide me through this idea?
I found the "Before Layer Change" Part in Orcaslicer, which contains
`;BEFORE_LAYER_CHANGE
;[layer_z]
G92 E0 ;`

So I just added the line you proposed, so it looks like this
`;BEFORE_LAYER_CHANGE
;[layer_z]
if layer_z == 2 SETFILAMENT_SENSOR SENSOR=SFS_T0 ENABLE=1`

I can slice with this setting but the printer send the error message that it does not know the command IF and it continues printing without switching on the sensor.
I did some googling on that and then changed to code to
`;BEFORE_LAYER_CHANGE
;[layer_z]
G92 E0 ;
{if layer_z == 2} SETFILAMENT_SENSOR SENSOR=SFS_T0 ENABLE=1
`
With this I cannot even slice as I am getting the error message that a tag literal-char is expected and the custom gcode is wrong.
So how do I aplly this idea then?

**Reply by foreachthing** [reply]
See here, for more info: https://help.prusa3d.com/article/macros_1775
Most of ir should work with OrcaSlicer as well...

**Reply by lullaby37** [reply]
Yes, that is the page I also found and that's why I added the "{ }" in my code as mihi's idea did not work.
But I am too bad at coding to know what the problem here is. As I said, the slicer says a tag literal-char is expected and therefore my custom gcode is wrong.

**Comment by mihi4** [comment]
hmmmm...  first, I think ";" starts a comment, so best to move the code in a seperate line
you have to set "layer_z" into brackets, otherwise the slicer would not see it as variable.
I haven't tried it, but this should work:

`{if [layer_z] == 2} SETFILAMENT_SENSOR SENSOR=SFS_T0 ENABLE=1`

**Reply by lullaby37** [reply]
Yes, I know the ";" starts the comment. Actually my comment has 3 lines, it's just the format here that makes it look like it is one line. Actually my code is the line you have there but it does not work.
I realized I had it in the "before layer change" box and not in the one for layer change itself. I changed that but it still does not work.

**Reply by mihi4** [reply]
So my line above did not work?
because it's different to yours, "layer_z" is between [ ] in my code but without in yours.

**Reply by lullaby37** [reply]
Oh, I did not see that, sorry.
I just tried and this one does not work either. This time the error message states that the slicer is exoecting a boolean expresion.

**Reply by lullaby37** [reply]
I tried to changer the example from the prusa website and came up with the following line
`{if layer_height > 1};SETFILAMENT_SENSOR SENSOR=SFS_T0 ENABLE=1{endif}`
I can now slice it and will test if the sensor is switched on once the ongoing print is finished

**Reply by mihi4** [reply]
I think it's the " {endif}" part that was missing. "== 2" should usually work
your example would call the macro on ALL layers after 1, so also on layer 3,4,5,6... up to the top

**Reply by mihi4** [reply, marked as answer]
I also found, that there should be a "layer_num" parameter, maybe use this instead of layer_z ?
`{if layer_num == 2};SET_FILAMENT_SENSOR SENSOR=SFS_T0 ENABLE=1{endif}`

**Reply by mihi4** [reply]
" [layer_z] is the current height of the nozzle in mm. [layer_num] is the layer currently being printed. " from
https://forum.prusa3d.com/forum/prusaslicer/show-layer-number-and-height-but-wrong/

**Reply by lullaby37** [reply]
It now seems to work, I marked your idea as the answer, although you might want to change the typo and add a "_" after set.
`{if layer_num == 2};SET_FILAMENT_SENSOR SENSOR=SFS_T0 ENABLE=1{endif}`
Thats the working code. I'll test and report.

**Reply by mihi4** [reply]
great that it worked... I copied the SETFILAMENT_SENSOR part from one of your posts above *gg* - looks like I got it wrong in my pseudocode-answer :-D

**Comment by abtblk** [comment]
Where is the layer change script section?

**Comment by mihi4** [comment]
Here's in the printer settings (Advanced mode)

![image](https://github.com/SoftFever/OrcaSlicer/assets/37828365/5bd2677b-6fd8-4eed-91fe-ab4b202de146)

**Comment by abtblk** [comment]
Thank you so much

**Comment by Testmangh** [comment]
I want to to add STOP_LED_EFFECTS to every 3rd layer but for some reason a blank line is being generated for the entire gcode, what can I do to fix?

![image](https://github.com/SoftFever/OrcaSlicer/assets/25848686/3c510787-f395-482b-9503-0d8f55cdd309)

![image](https://github.com/SoftFever/OrcaSlicer/assets/25848686/529f8bb5-a8b6-4720-8ee7-471d623a4709)

![image](https://github.com/SoftFever/OrcaSlicer/assets/25848686/28c63afe-78f8-4587-af7e-ea1fb3ec84f3)

**Reply by abtblk** [reply]
Does it do it on layer 3?

**Reply by abtblk** [reply]
Very weird I have no idea then

**Reply by djurny** [reply]
your conditional only checks if the `layer_num` equals `3`, which happens only once (or once for each object when printing per object). the conditional should check if the `layer_num` **modulo** `3` is 0, but i'm not sure if orcaslicer can process the modulo operator or not:
```
{if layer_num % 3 == 0}STOP_LED_EFFECTS{endif}
```
you can try it out and see if the above works for you.

ps the empty line is from orcaslicer resolving the `{if..}...{endif}` which resulted in an empty line.

**Comment by Testmangh** [comment]
> Does it do it on layer 3?
@TyTyDaCodingGuy  Yes > line 9227 on the first picture... But it leaves a blank line in every other layer > line 9914 on the second picture.

**Reply by Testmangh** [reply]
I think there is no problem, I'm coming from Cura and using search and replace it wouldn't leave a blank line, but there are other blank lines in the gcode file, so that should be fine, thanks...
